### DISCUSSION TITLE
raylib-go without cgo (dynamic loading)

### DISCUSSION THREAD

**Discussion by JupiterRider** [gen2brain/raylib-go#306] [Ideas]
With the `syscall` package in go's standard library or `golang.org/x/sys` it is already possible to dynamically load libraries without cgo. 

Works kinda like this: https://learn.microsoft.com/en-us/windows/win32/dlls/using-run-time-dynamic-linking

Here is a small example, of just displaying a white window:
```
.
├── go.mod
├── main.go
└── raylib.dll
```

```golang
package main

import (
	"image/color"
	"runtime"
	"syscall"
	"unsafe"
)

func init() {
	runtime.LockOSThread()
}

func main() {
	handle, _ := syscall.LoadLibrary("raylib.dll")
	defer syscall.FreeLibrary(handle)

	initWindow, _ := syscall.GetProcAddress(handle, "InitWindow")
	closeWindow, _ := syscall.GetProcAddress(handle, "CloseWindow")
	windowShouldClose, _ := syscall.GetProcAddress(handle, "WindowShouldClose")
	beginDrawing, _ := syscall.GetProcAddress(handle, "BeginDrawing")
	endDrawing, _ := syscall.GetProcAddress(handle, "EndDrawing")
	clearBackground, _ := syscall.GetProcAddress(handle, "ClearBackground")

	title, _ := syscall.BytePtrFromString("My Game")
	syscall.SyscallN(initWindow, uintptr(1280), uintptr(720), uintptr(unsafe.Pointer(title)))

	for {
		syscall.SyscallN(beginDrawing)
		syscall.SyscallN(clearBackground, *(*uintptr)(unsafe.Pointer(&color.RGBA{255, 255, 255, 255})))
		syscall.SyscallN(endDrawing)
		ret, _, _ := syscall.SyscallN(windowShouldClose)
		close := *(*bool)(unsafe.Pointer(&ret))
		if close {
			break
		}
	}

	syscall.SyscallN(closeWindow)
}
```

For Linux and MacOS there is a project called "purego": https://github.com/ebitengine/purego
But sadly structs are [not supported](https://github.com/ebitengine/purego/issues/153) in purego.
As a workaround you could write wrapper functions or try to get libffi working: https://github.com/ebitengine/purego/issues/88#issuecomment-1602098330

Using dynamic loading would reduce build times and enable cross compiling without any third party c-compiler. 

So what do you think?

**Comment by gen2brain** [comment]
The only viable option for removing cgo seems to be on Windows, where the code to load the dynamic library is in the kernel and not in the libC.

The purego without the structs looks like it would be a real pain. Anyway, if ever done, purego says it can be also used for Windows, so then I would not see a point in having a different code base just for Windows.

It would also remove so many options, we would need one .dll/.so file for every combination, i.e., one for SDL, one for DRM, Wayland, and what about OpenGL versions, etc., or if you want to override something in config.h, etc. Currently having C code included in the repo has many advantages. On the other hand, without cgo, I guess building for Web platform should be possible in theory, now that would be very nice.

There is also one more possible way, to compile raylib library to `raylib.wasm` with `wasi-sdk` and then run that with `wazero`. For example, see https://github.com/yklcs/wasmimg or https://github.com/wasilibs/go-re2. I haven't tried that method yet, but I probably will, with something useful, like libjxl.

**Reply by JupiterRider** [reply]
That's right - purego supports Windows as well. As far as I see, it uses `golang.org/x/sys` under the hood. 

The Idea is, that people just download the .dll from https://github.com/raysan5/raylib/releases and put it next to the executable/inside the project folder.

If we want to use this in raylib-go, we could separate that code by adding the `//go:build !cgo` tag.

**Reply by gen2brain** [reply]
That sounds great, also the part about `!cgo`. Now this just needs to be implemented :)

**Reply by gen2brain** [reply]
Btw. raylib for some time has https://github.com/raysan5/raylib/tree/master/parser with JSON output. That can be used to generate some code, at least for start. Doesn't have to be used later on.

**Reply by JupiterRider** [reply]
Is there some kind of code generator to convert the JSON into golang source code?

**Reply by gen2brain** [reply]
That is something that would need to be written, especially for purego.

**Reply by JupiterRider** [reply]
@gen2brain Do you recommend any source code generators for go? Jennifer looks pretty nice: https://github.com/dave/jennifer

**Reply by gen2brain** [reply]
I don't have experience with any of those.

**Comment by JupiterRider** [comment]
You can now use raylib-go without cgo on several platforms.
